cmd/argocd/commands/app.go

```diff
@@ -1799,13 +1821,51 @@ func NewApplicationWaitCommand(clientOpts *argocdclient.ClientOptions) *cobra.Co
 }
 
 // printAppResources prints the resources of an application in a tabwriter table
-func printAppResources(w io.Writer, app *argoappv1.Application) {
+func printAppResources(w io.Writer, app *argoappv1.Application, discoveryClient discovery.DiscoveryInterface) {
 	_, _ = fmt.Fprintf(w, "GROUP\tKIND\tNAMESPACE\tNAME\tSTATUS\tHEALTH\tHOOK\tMESSAGE\n")
-	for _, res := range getResourceStates(app, nil) {
+	versionedResourceStates := getVersionedResourceStates(app, nil)
+	for _, res := range mergeDuplicateClusterLevelResources(versionedResourceStates, discoveryClient) {
 		_, _ = fmt.Fprintf(w, "%s\t%s\t%s\t%s\t%s\t%s\t%s\t%s\n", res.Group, res.Kind, res.Namespace, res.Name, res.Status, res.Health, res.Hook, res.Message)
 	}
 }
 
+func mergeDuplicateClusterLevelResources(resourceStates []*versionedResourceState, discoveryClient discovery.DiscoveryInterface) []*versionedResourceState {
+	resourcesByKey := make(map[kube.ResourceKey]*versionedResourceState)
+	for _, resource := range resourceStates {
+		groupVersion := ""
+		if resource.Group != "" {
+			groupVersion += (resource.Group + "/")
+		}
+		groupVersion += resource.Version
+		apiResourceList, err := discoveryClient.ServerResourcesForGroupVersion(groupVersion)
```

Comment:
This call looks like it might be slow.

Reply:
Maybe, but this code path is only traversed from CLI commands, and how slow can it be? It boils down to few HTTP requests at most.